broker/lib/bosh/BoshOperationCache.js

```diff
@@ -0,0 +1,243 @@
+          out = out.map(v => v.substring(CACHE_KEY.length));
+        } else {
+          throw new Error("Unexpected output");
+        }
+        resolve(out);
+      }).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Fetches all BOSH deployment names in the Etcd store currently
+   */
+  getDeploymentNames() {
+    logger.debug('Getting the current deployment cache...');
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().getAll().prefix(CACHE_KEY).sort('Create', 'None').keys().then(out => {
+        if (Array.isArray(out)) {
+          out = out.map(v => v.substring(CACHE_KEY.length));
+        } else {
+          throw new Error("Unexpected output");
+        }
+        resolve(out);
+      }).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Fetches BOSH task ID for a service instance from the Etcd store
+   * 
+   * @param {Service instance ID} serviceInstanceId 
+   */
+  getBoshTask(serviceInstanceId) {
+    logger.debug(`Getting bosh task ID for service instance id ${serviceInstanceId}`);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getTaskKey(serviceInstanceId);
+      etcdConnector().get(key).string().then(obj => {
+        resolve(obj);
+      }).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Fetches parameters for a BOSH deployment operation from the Etcd store based on the name
+   * 
+   * @param {Deployment Name} name 
+   */
+  getDeploymentByName(name) {
+    logger.debug('Getting deployment for ', name);
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().get(getKey(name)).json().then(out => resolve(out)).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes a single BOSH deployment from the Etcd Store
+   * 
+   * @param {Deployment Name} deploymentName 
+   */
+  deleteDeploymentFromCache(deploymentName) {
+    logger.info('Removing deployment from cache', deploymentName);
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().delete().key(getKey(deploymentName)).then(out => resolve(out)).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes multiple BOSH deployment names from the Etcd store
+   * 
+   * @param {Deployment Names} deployments 
+   */
+  deleteDeploymentsFromCache(...deployments) {
+    const deploymentIdentifiers = _.flattenDeep(deployments);
+    logger.info('Removing multiple deployments from cache', deployments);
+    const deletePromises = _.map(deploymentIdentifiers, this.deleteDeploymentFromCache);
+    return Promise.all(deletePromises);
+  }
+
+  /**
+   * Stores the BOSH task ID corresponding to the service instance ID when the task is submitted to the BOSH Director
+   * 
+   * @param {Service Instance ID} serviceInstanceId 
+   * @param {BOSH Task ID corresponding to the operation} taskId 
+   */
+  storeBoshTask(serviceInstanceId, taskId) {
```

Comment:
Is this for the case where we do not want to hit the BOSH director?

Reply:
Yes...

Reply:
Why is that we have to maintain a task submitted to BOSH in the cache? The understanding is that the task which cannot be currently published to BOSH because of not enough capacity are kept in the cache

Reply:
The task ID needs to be put into Etcd because:
- If the request is staggered initially, the response `ACCEPTED` (202) would not contain the task ID
- CC will call the last operation endpoint of the broker repeatedly with the information (this needs to indicate that the request was cached)
- The poller would submit to BOSH at one point in time
- The subsequent last operation call must figure out the task ID and the only way to do so is via ETCD